keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -0,0 +1,212 @@
+package com.netflix.spinnaker.keel.sql
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.keel.api.ApiVersion
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceMetadata
+import com.netflix.spinnaker.keel.api.ResourceName
+import com.netflix.spinnaker.keel.persistence.NoSuchResourceException
+import com.netflix.spinnaker.keel.persistence.ResourceRepository
+import com.netflix.spinnaker.keel.persistence.ResourceState
+import com.netflix.spinnaker.keel.persistence.ResourceState.Unknown
+import com.netflix.spinnaker.keel.persistence.ResourceState.valueOf
+import de.huxhorn.sulky.ulid.ULID
+import org.jooq.DSLContext
+import org.jooq.exception.SQLDialectNotSupportedException
+import org.jooq.impl.DSL
+import org.jooq.impl.DSL.field
+import org.jooq.impl.DSL.using
+import org.slf4j.LoggerFactory
+import java.sql.ResultSet
+import java.sql.Timestamp
+import java.time.Clock
+import java.time.Instant
+
+class SqlResourceRepository(
+  private val jooq: DSLContext,
+  private val objectMapper: ObjectMapper,
+  private val clock: Clock
+) : ResourceRepository {
+
+  override fun allResources(callback: (Triple<ResourceName, ApiVersion, String>) -> Unit) {
+    jooq
+      .select(
+        field("api_version"),
+        field("kind"),
+        field("name")
+      )
+      .from(RESOURCE)
+      .fetch()
+      .intoResultSet()
+      .apply {
+        while (next()) {
+          callback(Triple(resourceName, apiVersion, kind))
+        }
+      }
+  }
+
+  override fun <T : Any> get(name: ResourceName, specType: Class<T>): Resource<T> {
+    return jooq
+      .select(
+        field("uid"),
+        field("api_version"),
+        field("kind"),
+        field("name"),
+        field("resource_version"),
+        field("spec")
+      )
+      .from(RESOURCE)
+      .where(field("name").eq(name.value))
+      .fetch()
+      .intoResultSet()
+      .run {
+        if (next()) {
+          toResource(specType)
+        } else {
+          throw NoSuchResourceException(name)
```

Comment:
Interested to know why you prefer throwing exceptions for not found instead of a nullable type?

Reply:
In this case it’s a truly exceptional condition. There’s no reason this method should ever get called with an invalid name.